storage/aptosdb/src/lib.rs

```diff
@@ -1432,35 +1443,71 @@ impl DbWriter for AptosDB {
                 self.commit(sealed_cs)?;
             }
             if save_state_snapshots {
-                let checkpoint_version =
+                let cached_snapshot_version =
                     self.state_store.in_memory_state().lock().checkpoint_version;
-                let mut base_version = base_state_version;
                 ensure!(
-                    base_version == checkpoint_version,
-                    "base_version {:?} does not match checkpoint_version {:?} in state_store",
-                    base_version,
-                    checkpoint_version
+                    base_state_version == cached_snapshot_version,
+                    "base_state_version {:?} does not match cached_snapshot_version {:?} in state_store",
+                    base_state_version,
+                    cached_snapshot_version
                 );
-                // find all the checkpoint versions
-                for (idx, txn_to_commit) in txns_to_commit.iter().enumerate() {
+
+                let mut uncommitted_updates_start_idx = 0;
+                let new_snapshot_version = latest_in_memory_state.checkpoint_version;
+                if new_snapshot_version >= Some(first_version) {
+                    let new_snapshot_version = new_snapshot_version.expect("Must exist here");
+                    ensure!(
+                        last_version >= new_snapshot_version,
+                        "last_version is smaller than new_snapshot version {} {}",
+                        last_version,
+                        new_snapshot_version
+                    );
+                    let idx = (new_snapshot_version - first_version) as usize;
+                    ensure!(
+                            txns_to_commit[idx].is_state_checkpoint(),
+                            "The new snapshot version passed in {:?} does not match with the last checkpoint version in txns_to_commit {:?}",
+                            new_snapshot_version,
+                            first_version + idx as u64
+                    );
                     self.state_store
                         .in_memory_state()
                         .lock()
                         .updated_since_checkpoint
-                        .extend(txn_to_commit.state_updates().clone());
-                    if txn_to_commit.is_state_checkpoint() {
-                        let version = first_version + idx as LeafCount;
-                        self.save_state_snapshot(version, base_version, state_tree.clone())?;
-                        base_version = Some(version);
+                        .extend(
```

Comment:
how did we make sure the last version of current cache is `first_version - 1`?

Reply:
We can verify that